### DISCUSSION TITLE
From Seurat to AnnData and scVelo

### DISCUSSION THREAD

**Discussion by Jeff87075** [theislab/scvelo#452] [Q&A] [answered]
Hi, apologies for such a basic question but I'm pretty confused about the input files required for running the ```scvelo``` analysis workflow. I am familiar with Seurat but am new to python and the ultimate objective is to run the RNA velocity analysis but with the UMAP embedding and clustering I've processed in the Seurat object.

I've been trying to go through the tutorial and know that both the adata and ldata should be loom files but I'm uncertain as to what exactly is the difference between the two and how/where they can be generated. Any guidance would be appreciated, thanks!

**Comment by WeilerP** [comment, marked as answer]
@Jeff87075, sorry, not sure what you mean by `ldata`. Could you elaborate?

I'll try to briefly outline how to proceed on going from a Seurat object to AnnData and using scVelo:

### Conversion from Seurat to AnnData

When you have your Seurat object `data` you can, for example, convert it to `AnnData` using `anndata2ri`. See, for example, [here](https://github.com/LuckyMD/Code_snippets/blob/master/Seurat_to_anndata.ipynb) how you can use it in a Jupyter notebook. This will keep the UMAP embedding (should be added as `adata.obsm['X_umap']`) and clusters (stored as column in `adata.obs`).

Alternatively, if you want to do everything in R, you can use `Seurat` and `SeuratDisk`:

```R
library(Seurat)
library(SeuratDisk)

data <- ReadRDS(data.rds)
SaveH5Seurat(data, filename = "data_from_seurat.h5Seurat")
Convert('data_from_seurat.h5Seurat', dest='h5ad')
```

**Disclaimer:** I previously ran into problems using the latter approach when the neighbor graph shape did not match the count matrix (e.g. more nodes than observations because they were subsetted after graph construction).

### Loading loom files and aligning data

```python
import scanpy as sc

adata_from_seurat = sc.read()
adata = sc.read_loom(LOOM_FILE.loom)

# Remove observations/features not present in data from Seurat
adata = adata[
    adata_from_seurat.obs_names.intersection(adata.obs_names),
    adata_from_seurat.var_names.intersection(adata.var_names)
]

# Add Seurat UMAP embedding and clusters
adata.obs['seurat_clusters'] = adata_from_seurat.obs['seurat_clusters']
adata.obsm['X_umap'] = adata_from_seurat.obsm['X_umap']
```

**Reply by VolkerBergen** [reply]
`ldata`, as used in line 4 in the [tutorial](https://scvelo.readthedocs.io/getting_started/), is the loom file generated by velocyto, while `adata` is the object processed by Seurat. `scv.utils.merge(adata, ldata)` can be used to merge the two (un/spliced counts from loom, and embedding coord. from Seurat).

**Reply by WeilerP** [reply]
Ah yes, thanks @VolkerBergen. `scv.utils.merge` shorter and more elegant.

**Reply by Jeff87075** [reply]
Thanks for the clear explanations. @VolkerBergen I actually have the ```ldata``` files now so I just have to figure out how to obtain the ```adata``` loom file and I think I have got it figured out using the ```as.loom()``` function. But I was wondering then what's the difference/relationship between scvelo and velocyto? Or in other words is the output files from velocyto an input file for scvelo and in the future if I do not have the files from velocyto to begin with, do I just skip the ```scv.utils.merge(adata,ldata)``` step and follow the workflow as stated by @WeilerP ? 

Also one more thing I'd like to clarify @WeilerP for this step:
```
adata_from_seurat = sc.read()
adata = sc.read_loom(LOOM_FILE.loom)
```
If I'm understanding it correctly, the ```adata_from_seurat``` is the ```AnnData``` file converted from Seurat but then where does the loom file for ```adata``` come from? Many thanks.

**Reply by WeilerP** [reply]
Yes, `adata_from_seurat` is the `AnnData` converted from Seurat. The loom file comes from running the velocyto pipeline `velocyto run ...`. This also answers your other question: The output of the velocyto pipeline, i.e. the loom file, is used as input for `scvelo`. It provides the `unspliced`, `spliced` count matrices used to calculate RNA velocity.

The second part of the workflow I outlined is almost the same as the other one. The difference being that you align the `AnnData` objects yourself and manually pick slots and columns to add.

**Reply by Jeff87075** [reply]
I see, sorry just to make it very clear, so can I take it as the output loom file (with the unspliced/spliced count matrix) from velocyto pipeline is a must/prerequisite for running scvelo? Also is it the ```adata_from_seurat``` can also be a loom file instead of the ```AnnData``` H5AD file type? I'm asking this because the conversion to ```AnnData``` using the ```SeuratDisk``` functions seems to be giving some errors. Thanks a lot.

**Reply by WeilerP** [reply]
Yes, unspliced and spliced count matrices are needed for running `scvelo`. They are the basis of [RNA velocity](https://www.nature.com/articles/s41586-018-0414-6/) in general!

And yes, `adata_from_seurat` can be read as a loom file as well. In general, `scanpy` supports reading a vast number of file types into an `AnnData` object. See [here](https://scanpy.readthedocs.io/en/stable/api/index.html#reading) for the official documentation.

**Reply by WeilerP** [reply]
@Jeff87075, did this answer your question? If so, would you please mark the discussion as answered?
